src/simulators/matrix_product_state/matrix_product_state_internal.cpp

```diff
@@ -1481,24 +1481,37 @@ reg_t MPS::apply_measure(const reg_t &qubits, RngEngine &rng) {
 
 reg_t MPS::apply_measure_internal(const reg_t &qubits, 
 				  RngEngine &rng) {
-  // When all qubits are measured, then for every qubit measured, it is sufficient to 
-  // propagate to the nearest neighbors because the neighbors will be measured next
-  bool measure_all = 0;
-  if (qubits.size() == num_qubits_)
-    measure_all = true;
+  // For every qubit, q,  that is measured, we must propagate the effect of its
+  // measurement to its neigbors, l and r, and then to their neighbors, and 
+  // so on. If r (or l) is measured next, then there is no need to propagate to
+  // its next neighbor because we can propagate the effects of measuring q 
+  // and r together.
+  // We sort 'qubits' at the beginning of the algorithm, so that the index of 
+  // the next measured qubit will always be r, or greater. Therefore we check 
+  // if r needs to be measured. If so, we simply measure it. If not, we 
+  // propagate the effect of measuring q all the way to the right. 
+  // In both cases, we propagate the effect all the way to the left.
   reg_t qubits_to_update;
-  reg_t outcome_vector(qubits.size());
-  for (uint_t i=0; i<qubits.size(); i++) {
+  uint_t size = qubits.size();
+  reg_t outcome_vector(size);
+  reg_t sorted_qubits = qubits;
+  std::sort(sorted_qubits.begin(), sorted_qubits.end());
+
+  bool measure_right_neighbor = false;
+  // here we check if the next qubit to be measured is r (r's index is i+1)
+  for (uint_t i=0; i<size; i++) {
+    measure_right_neighbor = (i<size-1 && sorted_qubits[i+1] == i+1);
```

Comment:
```
measure_right_neighbor = (i<size-1 && sorted_qubits[i] + 1 == sorted_qubits[i+1]);
```
Please add tests

Reply:
Fixed bug, as you pointed out and added tests.

Reply:
I think perhaps you forgot to push